pkg/liqoctl/test/client/k8s.go

```diff
@@ -0,0 +1,125 @@
+
+package client
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/liqotech/liqo/pkg/liqoctl/test"
+)
+
+// Client is a struct that contains all the k8s clients used in tests.
+type Client struct {
+	Consumer        client.Client
+	ConsumerName    string
+	ConsumerDynamic *dynamic.DynamicClient
+	// Providers key is the name of the producer kubeconfig file.
+	Providers map[string]client.Client
+	// ProvidersDynamic key is the name of the producer kubeconfig file.
+	ProvidersDynamic map[string]*dynamic.DynamicClient
+}
+
+// Configs is a map that contains the rest configuration for each client.
+type Configs map[string]*rest.Config
+
+// NewClient returns a new Client struct.
+func NewClient(ctx context.Context, opts *test.Options) (*Client, Configs, error) {
+	var cl Client
+	cfg := Configs{}
+
+	if err := initConfigAndClient(ctx, "", &cl, cfg, opts); err != nil {
+		return nil, nil, err
+	}
+
+	for _, kubeconfig := range opts.RemoteKubeconfigs {
+		if err := initConfigAndClient(ctx, kubeconfig, &cl, cfg, opts); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return &cl, cfg, nil
+}
+
+func initConfigAndClient(ctx context.Context, kubeconfig string, cl *Client, cfg Configs, opts *test.Options) error {
+	var err error
+	var cfgtmp *rest.Config
+	var cltmp client.Client
+	var cldyntmp *dynamic.DynamicClient
+	if kubeconfig == "" {
+		cfgtmp = opts.LocalFactory.RESTConfig
+		cltmp = opts.LocalFactory.CRClient
+		cldyntmp, err = dynamic.NewForConfig(cfgtmp)
+		if err != nil {
+			return err
+		}
+	} else {
+		cfgtmp, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			return err
+		}
+		cltmp, err = client.New(cfgtmp, client.Options{
+			Scheme: opts.LocalFactory.CRClient.Scheme(),
+		})
+		if err != nil {
+			return err
+		}
+		cldyntmp, err = dynamic.NewForConfig(cfgtmp)
+		if err != nil {
+			return err
+		}
+	}
+
+	name, err := getClientName(ctx, cltmp)
+	if err != nil {
+		return err
+	}
+
+	cfg[name] = cfgtmp
+	if cl.ConsumerName == "" {
+		cl.ConsumerName = name
+		cl.Consumer = cltmp
+		cl.ConsumerDynamic = cldyntmp
+	} else {
+		if cl.Providers == nil {
+			cl.Providers = make(map[string]client.Client)
+		}
+		cl.Providers[name] = cltmp
+
+		if cl.ProvidersDynamic == nil {
+			cl.ProvidersDynamic = make(map[string]*dynamic.DynamicClient)
+		}
+		cl.ProvidersDynamic[name] = cldyntmp
+	}
+
+	return err
+}
+
+func getClientName(ctx context.Context, cl client.Client) (string, error) {
+	cfgmap := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{Name: "liqo-clusterid-configmap", Namespace: "liqo"},
+	}
+	if err := (cl).Get(ctx, client.ObjectKeyFromObject(&cfgmap), &cfgmap); err != nil {
+		return "", fmt.Errorf("failed to get cluster ID: %w", err)
+	}
+	return cfgmap.Data["CLUSTER_ID"], nil
+}
```

Comment:
Why not using `GetClusterIDWithControllerClient()` function?

Reply:
Because I didn't know it was there :)